### PR TITLE
feat(chmod): add Linux file permissions calculator tool (closes #82)

### LINKED ISSUE #82: feat: Linux chmod Permissions Calculator — interactive visual matrix
## Summary
Add a **Linux File Permissions (chmod) Calculator** tool. One of the single most Googled Linux queries — developers rarely memorize the numeric weights for all permission combinations and special bits.

## Proposed Implementation
- **Location:** `src/tools/chmod/ChmodTool.tsx`
- **Zero dependencies** — trivial bitwise math
- **Features:**
  - Interactive permission matrix with checkboxes for User, Group, Other
  - Permissions: Read (4), Write (2), Execute (1)
  - Special bits: SUID (4000), SGID (2000), Sticky Bit (1000)
  - Live outputs:
    - Octal notation (`755`, `644`, `600`)
    - Symbolic notation (`rwxr-xr-x`, `rw-r--r--`)
    - Generated `chmod` command (`chmod 755 filename`)
    - `ls -la` style display
  - **Common security presets:**
    - SSH Private Key: `600` (`-rw-------`)
    - Web Server Directory: `755` (`drwxr-xr-x`)
    - Executable Script: `754` (`-rwxr-xr--`)
    - Public Read-Only: `644` (`-rw-r--r--`)
    - Temp Directory (Sticky): `1777` (`drwxrwxrwt`)
  - Reverse lookup: input octal → show checkboxes
  - Copy any output format

## Acceptance Criteria
- [ ] All permission combinations calculate correctly
- [ ] Special bits (SUID/SGID/Sticky) work
- [ ] Reverse lookup (octal → visual) works
- [ ] Presets load correctly
- [ ] Zero dependencies

### PR DESCRIPTION
**Pull request by ajithakdev**
## Summary

Implements the **Linux chmod Permissions Calculator** tool requested in #82.

### Features
- **Interactive Matrix**: Checkboxes for User (Owner), Group, and Other across Read (4), Write (2), Execute (1).
- **Special Bits Support**: Toggle SUID (4000), SGID (2000), and Sticky Bit (1000) with capital/lowercase symbolic flags (`s`, `S`, `t`, `T`).
- **Live Output Formats**:
  - Octal representation (e.g. `755`, `644`, `1777`)
  - Symbolic notation (e.g. `rwxr-xr-x`, `rwxrwxrwt`)
  - Full shell command (e.g. `chmod 755 filename.sh`) with customizable file name
  - Symbolic command (e.g. `chmod u=rwx,g=rx,o=rx filename.sh`)
  - `ls -la` style output preview
- **Quick Security Presets**: One-click application of common permission configs (SSH key 600, Standard file 644, Executable 755, Sticky temp 1777, Full 777, etc.).
- **Reverse Lookup**: Type any 3- or 4-digit octal value into the input to instantly populate the visual matrix.
- **URL Synchronization**: Share exact permission configurations via URL parameter (`?v=755`).
- **Zero External Dependencies**: Pure bitwise math and client-side React.

### Testing
- **17 new unit & component tests**:
  - 12 unit tests in `chmod.test.ts` verifying all octal math, symbolic logic, special bits, presets, and error edge cases.
  - 5 component tests in `chmodTool.test.tsx` covering interactive checkbox changes, preset buttons, reverse text input, and reset.
- `npm run lint` passes with 0 warnings / 0 errors.
- `npm run build` compiles cleanly in ~700ms.

Closes #82

---

<a href="https://app.devin.ai/review/ajithakdev/toolglass/pull/93" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

File: src/tools/chmod/chmod.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+  const u = getPermStr(state.user) + (state.special.suid ? 's' : '');
+  const g = getPermStr(state.group) + (state.special.sgid ? 's' : '');
+  const o = getPermStr(state.other) + (state.special.sticky ? 't' : '');
+
+  if (u) parts.push(`u=${u}`);
+  if (g) parts.push(`g=${g}`);
+  if (o) parts.push(`o=${o}`);
+
+  return parts.length > 0 ? parts.join(',') : 'ugo=';
```
[the remark this answers is not in the document]
Comment (ajithakdev) [reply src/tools/chmod/chmod.ts]: Fixed in commit d686146.  now returns explicit clauses for all classes () so empty classes are cleared as expected by chmod.

File: src/tools/chmod/ChmodTool.tsx [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+            }}
+          >
+            <div style={{ fontSize: 12, fontWeight: 600, color: 'var(--ink-mute)', textTransform: 'uppercase' }}>
+              Octal Notation
+            </div>
+            <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between' }}>
+              <input
+                type="text"
+                value={octal}
+                onChange={(e) => handleOctalInput(e.target.value)}
```
[the remark this answers is not in the document]
Comment (ajithakdev) [reply src/tools/chmod/ChmodTool.tsx]: Addressed in commit d686146. Separated `octalInput` editable string state from the canonical `ChmodState`, allowing fluid intermediate 1-3 digit editing without early rejection, updating state as soon as valid 3-4 digits are entered, and resetting on blur if incomplete.

File: src/tools/chmod/ChmodTool.tsx [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+  const reset = () => {
+    setState(DEFAULT_CHMOD_STATE);
+    setOctalParam('755');
+  };
+
+  const chmodCmd = `chmod ${octal} ${filename}`;
+  const chmodSymCmd = `chmod ${symbolicCmd} ${filename}`;
+  const isDirectory = state.special.sticky || octal.endsWith('5');
+  const lsStyle = `${isDirectory ? 'd' : '-'}${symbolic} 1 user staff 4096 Sep 7 12:00 ${filename}`;
```
[the remark this answers is not in the document]
Comment (ajithakdev) [reply src/tools/chmod/ChmodTool.tsx]: Addressed in commit d686146. Replaced the permission heuristic with an explicit File / Directory segmented toggle beside the target name, setting the standard `d` or `-` prefix accordingly in the ls preview.

File: src/tools/chmod/ChmodTool.tsx [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    }
+  };
+
+  const reset = () => {
+    setState(DEFAULT_CHMOD_STATE);
+    setOctalParam('755');
+  };
+
+  const chmodCmd = `chmod ${octal} ${filename}`;
+  const chmodSymCmd = `chmod ${symbolicCmd} ${filename}`;
```
[the remark this answers is not in the document]
Comment (ajithakdev) [reply src/tools/chmod/ChmodTool.tsx]: Resolved in commit d686146. Added POSIX-compliant `escapeShellArg` utility to sanitize filename input and prevent shell argument injection when generating terminal commands.

File: src/tools/chmod/ChmodTool.tsx [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+  type PermissionState,
+  type SpecialBitsState,
+} from './chmod';
+
+export default function ChmodTool() {
+  const [octalParam, setOctalParam] = useUrlState('v', '755', String, String);
+  const [filename, setFilename] = useState('script.sh');
+  const [fileType, setFileType] = useState<'file' | 'dir'>('file');
+
+  const [state, setState] = useState<ChmodState>(() => {
```
[the remark this answers is not in the document]
Comment (ajithakdev) [reply src/tools/chmod/ChmodTool.tsx:24]: Resolved in commit 13f79bb. Added a reactive `useEffect` on `octalParam` that parses and syncs valid URL parameter updates directly into `state`, enabling full back/forward history navigation and query parameter deep linking.

File: src/tools/chmod/chmodTool.test.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    ) as HTMLInputElement[];
+    const firstCheckbox = checkboxes[0]; // User read
+
+    await act(async () => {
+      firstCheckbox.click();
+      await new Promise((r) => setTimeout(r, 50));
+    });
+
+    expect(container.textContent).toContain('chmod');
+  });
```
[the remark this answers is not in the document]
Comment (ajithakdev) [reply src/tools/chmod/chmodTool.test.tsx:107]: Resolved in commit 13f79bb. Strengthened the checkbox toggle test to assert specific permission bit flipping (User Read toggling 755 to 355 `-wxr-xr-x`, updating symbolic command to `u=wx,g=rx,o=rx`, and restoring on second click), and added a dedicated test verifying URL query parameter parsing.
